storage/bootstrap/bootstrapper/fs/source.go

```diff
@@ -454,76 +621,81 @@ func (s *fileSystemSource) loadShardReadersDataIntoShardResult(
 		}
 	}
 
-	s.handleErrorsAndUnfulfilled(
-		resultLock, bootstrapResult, shard, tr, shardResult, timesWithErrors)
+	s.handleErrorsAndUnfulfilled(resultLock, runResult, shard,
+		tr, shardResult, timesWithErrors)
 }
 
-func (s *fileSystemSource) ReadData(
+func (s *fileSystemSource) read(
 	md namespace.Metadata,
 	shardsTimeRanges result.ShardTimeRanges,
-	_ bootstrap.RunOptions,
-) (result.DataBootstrapResult, error) {
+	run runType,
+) (*runResult, error) {
+	var (
+		nsID           = md.ID()
+		blockRetriever block.DatabaseBlockRetriever
+	)
 	if shardsTimeRanges.IsEmpty() {
-		return result.NewDataBootstrapResult(), nil
+		return newRunResult(), nil
 	}
 
-	var (
-		nsID              = md.ID()
-		blockRetrieverMgr = s.opts.DatabaseBlockRetrieverManager()
-		blockRetriever    block.DatabaseBlockRetriever
-	)
-	if blockRetrieverMgr != nil {
-		s.log.WithFields(
-			xlog.NewField("namespace", nsID.String()),
-		).Infof("filesystem bootstrapper resolving block retriever")
+	if run == bootstrapDataRunType {
+		// NB(r): We can only need to cache shard indices and possibly shortcut
```

Comment:
This looks like one of the few big code blocks that would be easy to factor out into a method since it doesn't modify any other function state except the blockRetriever (which it could return).

Also, its ok that the blockRetriever is always nil in the index case yeah?

Reply:
This comment is a little confusing as well.

Maybe something like: "We only need to cache shard indices and marks blocks as fulfilled when bootstrapping data, because the data can be retrieved lazily from disk during reads. On the other hand, if we're bootstrapping the index then we need to rebuild it from scratch by reading all the IDs/tags"